File: src/backend/pipeline_builder.py
# SPDX-FileCopyrightText: Copyright 2021 SeaDve
# SPDX-License-Identifier: GPL-3.0-or-later

from gi.repository import Gst, GLib

ENCODING_PROFILES = {
    'webm': {
        'muxer': 'webmmux',
        'video_enc': 'vp8enc max_quantizer=17 cpu-used=16 cq_level=13 deadline=1 '
                     'static-threshold=100 keyframe-mode=disabled buffer-size=20000 threads=%T',
        'audio_enc': 'opusenc',
    },
    'mkv': {
        'muxer': 'matroskamux',
        'video_enc': 'vp8enc max_quantizer=17 cpu-used=16 cq_level=13 deadline=1 '
                     'static-threshold=100 keyframe-mode=disabled buffer-size=20000 threads=%T',
        'audio_enc': 'opusenc',
    },
    'mp4': {
        'muxer': 'mp4mux',
        'video_enc': 'x264enc qp-max=17 speed-preset=superfast threads=%T '
                     '! video/x-h264, profile=baseline',
        'audio_enc': 'opusenc',
    },
    'gif': {
        'muxer': None,
        'video_enc': 'gifenc qos=true'
    },
}


class PipelineBuilder:

    def __init__(self, pipewire_stream):
        self.fd = pipewire_stream.fd
        self.node_id = pipewire_stream.node_id
        self.stream_screen = pipewire_stream.screen
        self.speaker_source = None
        self.mic_source = None
        self.coordinates = None

    def _even_out(self, *numbers):
        return (int(number // 2 * 2) for number in numbers)

    def _rescale(self, coordinates):
        scale_factor = self.stream_screen.w / self.actual_screen.w
        return (coordinate * scale_factor for coordinate in coordinates)

    def _get_thread_count(self):
        num_processors = GLib.get_num_processors()
        return min(num_processors, 64)

    def _get_video_enc(self):
        return ENCODING_PROFILES[self.video_format]['video_enc']

    def _get_audio_enc(self):
        return ENCODING_PROFILES[self.video_format]['audio_enc']

    def _get_muxer(self):
        muxer_name = ENCODING_PROFILES[self.video_format]['muxer']
        if not muxer_name:
            return None

        return f'{muxer_name} name=mux'

    def _get_scaler(self):
        if not self.coordinates:
            return None

        width = self.stream_screen.w
        height = self.stream_screen.h

        return f'videoscale ! video/x-raw, width={width}, height={height}'

    def _get_cropper(self):
        if not self.coordinates:
            return None

        x, y, width, height = self._rescale(self.coordinates)
        right_crop = (self.stream_screen.w - (width + x))
        bottom_crop = (self.stream_screen.h - (height + y))
        x, y, right_crop, bottom_crop = self._even_out(x, y, right_crop, bottom_crop)

        return f'videocrop top={y} left={x} right={right_crop} bottom={bottom_crop}'

    def _get_proper_framerate(self):
        if self.video_format == 'gif':
            return 15

        return self.framerate

    def _get_audio_configuration(self):
        if self.video_format == 'gif':
            return False, False

        is_record_speaker = self.audio_source_type.record_speaker and self.speaker_source
        is_record_mic = self.audio_source_type.record_mic and self.mic_source
        return is_record_speaker, is_record_mic

    def set_settings(self, framerate, file_path, video_format, audio_source_type):
        self.framerate = framerate
        self.file_path = file_path
        self.video_format = video_format
        self.audio_source_type = audio_source_type

    def set_audio_source(self, speaker_source, mic_source):
        self.speaker_source = speaker_source
        self.mic_source = mic_source

    def set_coordinates(self, coordinates, actual_screen):
        self.coordinates = coordinates
        self.actual_screen = actual_screen

    def build(self):
        pipeline_elements = [
            f'pipewiresrc fd={self.fd} path={self.node_id} do-timestamp=true keepalive-time=1000 resend-last=true',  # noqa: E501
            f'video/x-raw, max-framerate={self._get_proper_framerate()}/1',
            self._get_scaler(),
            'videorate',
            f'video/x-raw, framerate={self._get_proper_framerate()}/1',
            'videoconvert chroma-mode=GST_VIDEO_CHROMA_MODE_NONE dither=GST_VIDEO_DITHER_NONE matrix-mode=GST_VIDEO_MATRIX_MODE_OUTPUT_ONLY n-threads=%T',  # noqa: E501
            self._get_cropper(),
            'queue',
            self._get_video_enc(),
            'queue',
            self._get_muxer(),
            f'filesink location={self.file_path}'
        ]
        pipeline_string = ' ! '.join(filter(None, pipeline_elements))

        is_record_speaker, is_record_mic = self._get_audio_configuration()
        if is_record_speaker and is_record_mic:
            pipeline_string = f'{pipeline_string} pulsesrc device="{self.mic_source}" ! queue ! audiomixer name=mix ! {self._get_audio_enc()} ! queue ! mux. pulsesrc device="{self.speaker_source}" ! queue ! mix.'  # noqa: E501
        elif is_record_speaker:
            pipeline_string = f'{pipeline_string} pulsesrc device="{self.speaker_source}" ! {self._get_audio_enc()} ! queue ! mux.'  # noqa: E501
        elif is_record_mic:
            pipeline_string = f'{pipeline_string} pulsesrc device="{self.mic_source}" ! {self._get_audio_enc()} ! queue ! mux.'  # noqa: E501

        pipeline_string = pipeline_string.replace('%T', str(self._get_thread_count()))
        return Gst.parse_launch(pipeline_string)

File: src/backend/pipeline_builder.rs
use ashpd::desktop::screencast::Stream;
use gtk::glib;

use std::{
    env,
    path::{Path, PathBuf},
    str::FromStr,
};

use crate::{
    data_types::{Rectangle, Screen},
    utils,
};

const GIF_DEFAULT_FRAMERATE: u32 = 15;

#[derive(Debug, PartialEq, strum_macros::EnumString)]
#[strum(serialize_all = "snake_case")]
enum VideoFormat {
    Webm,
    Mkv,
    Mp4,
    Gif,
}

#[derive(Debug, Default, Clone)]
pub struct PipelineBuilder {
    is_record_speaker: bool,
    is_record_mic: bool,
    framerate: u32,
    file_path: PathBuf,
    fd: i32,
    streams: Vec<Stream>,
    speaker_source: Option<String>,
    mic_source: Option<String>,
    coordinates: Option<Rectangle>,
    actual_screen: Option<Screen>,
}

impl PipelineBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn record_speaker(mut self, is_record_speaker: bool) -> Self {
        self.is_record_speaker = is_record_speaker;
        self
    }

    pub fn record_mic(mut self, is_record_mic: bool) -> Self {
        self.is_record_mic = is_record_mic;
        self
    }

    pub fn framerate(mut self, framerate: u32) -> Self {
        self.framerate = framerate;
        self
    }

    pub fn file_path(mut self, file_path: PathBuf) -> Self {
        self.file_path = file_path;
        self
    }

    pub fn fd(mut self, fd: i32) -> Self {
        self.fd = fd;
        self
    }

    pub fn streams(mut self, streams: Vec<Stream>) -> Self {
        self.streams = streams;
        self
    }

    pub fn speaker_source(mut self, speaker_source: Option<String>) -> Self {
        self.speaker_source = speaker_source;
        self
    }

    pub fn mic_source(mut self, mic_source: Option<String>) -> Self {
        self.mic_source = mic_source;
        self
    }

    pub fn coordinates(mut self, coordinates: Rectangle) -> Self {
        self.coordinates = Some(coordinates);
        self
    }

    pub fn actual_screen(mut self, actual_screen: Screen) -> Self {
        self.actual_screen = Some(actual_screen);
        self
    }

    pub fn parse_into_string(self) -> String {
        PipelineParser::from_builder(self).parse()
    }

    pub fn build(self) -> Result<gst::Element, glib::Error> {
        let pipeline_string = self.parse_into_string();
        log::debug!("pipeline_string: {}", &pipeline_string);

        gst::parse_launch(&pipeline_string)
    }
}

struct PipelineParser {
    builder: PipelineBuilder,
}

impl PipelineParser {
    pub fn from_builder(builder: PipelineBuilder) -> Self {
        Self { builder }
    }

    pub fn parse(&self) -> String {
        let pipeline_elements = vec![
            self.compositor(),
            Some("queue name=queue0".to_string()),
            Some("videorate".to_string()),
            Some(format!("video/x-raw, framerate={}/1", self.framerate())),
            self.videoscale(),
            self.videocrop(),
            Some("videoconvert chroma-mode=GST_VIDEO_CHROMA_MODE_NONE dither=GST_VIDEO_DITHER_NONE matrix-mode=GST_VIDEO_MATRIX_MODE_OUTPUT_ONLY n-threads=%T".to_string()),
            Some("queue".to_string()),
            Some(self.videoenc()),
            Some("queue".to_string()),
            self.muxer(),
            Some(format!("filesink name=filesink location=\"{}\"", self.file_path().display())),
        ];

        let mut pipeline_string = pipeline_elements
            .into_iter()
            .flatten()
            .collect::<Vec<String>>()
            .join(" ! ");

        pipeline_string = format!(
            "{} {} {}",
            pipeline_string,
            self.pipewiresrc(),
            self.pulsesrc()
        );

        pipeline_string.replace("%T", &utils::ideal_thread_count().to_string())
    }

    fn compositor(&self) -> Option<String> {
        if self.is_single_stream() {
            return None;
        }

        let mut current_pos = 0;
        let mut compositor_elements = vec!["compositor name=comp".to_string()];

        for (sink_num, stream) in self.streams().iter().enumerate() {
            // This allows us to place the videos size by size with each other, without overlaps.
            let pad = format!("sink_{}::xpos={}", sink_num, current_pos);
            compositor_elements.push(pad);

            let stream_width = stream.size().unwrap().0;
            current_pos += stream_width;
        }

        Some(compositor_elements.join(" "))
    }

    fn pipewiresrc(&self) -> String {
        if self.is_single_stream() {
            let node_id = self.streams()[0].pipe_wire_node_id();

            // If there is a single stream, connect pipewiresrc directly to queue0.
            return format!("pipewiresrc fd={} path={} do-timestamp=true keepalive-time=1000 resend-last=true ! video/x-raw, max-framerate={}/1 ! queue0.", self.fd(), node_id, self.framerate());
        }

        let mut pipewiresrc_list = Vec::new();
        for stream in self.streams().iter() {
            let node_id = stream.pipe_wire_node_id();
            pipewiresrc_list.push(format!("pipewiresrc fd={} path={} do-timestamp=true keepalive-time=1000 resend-last=true ! video/x-raw, max-framerate={}/1 ! comp.", self.fd(), node_id, self.framerate()));
        }

        pipewiresrc_list.join(" ")
    }

    fn pulsesrc(&self) -> String {
        if self.video_format() == VideoFormat::Gif {
            return "".to_string();
        }

        let speaker_source = self.speaker_source();
        let mic_source = self.mic_source();

        let is_record_speaker = self.builder.is_record_speaker && speaker_source.is_some();
        let is_record_mic = self.builder.is_record_mic && mic_source.is_some();

        let audioenc = self.audioenc().unwrap();

        match (is_record_speaker, is_record_mic) {
            (true, true) => {
                format!("pulsesrc device=\"{}\" ! queue ! audiomixer name=mix ! {} ! queue ! mux. pulsesrc device=\"{}\" ! queue ! mix.",
                    speaker_source.unwrap(),
                    audioenc,
                    mic_source.unwrap()
                )
            }
            (true, false) => {
                format!(
                    "pulsesrc device=\"{}\" ! {} ! queue ! mux.",
                    speaker_source.unwrap(),
                    audioenc
                )
            }
            (false, true) => {
                format!(
                    "pulsesrc device=\"{}\" ! {} ! queue ! mux.",
                    mic_source.unwrap(),
                    audioenc
                )
            }
            (false, false) => "".to_string(),
        }
    }

    fn videoscale(&self) -> Option<String> {
        if self.builder.coordinates.is_some() {
            // We could freely get the first stream because screencast portal won't allow multiple
            // sources selection if it is selection mode. Thus, there will be always single stream
            // present when we have coordinates. (The same applies with videocrop).
            let stream = &self.streams()[0];
            let width = stream.size().unwrap().0;
            let height = stream.size().unwrap().1;

            Some(format!(
                "videoscale ! video/x-raw, width={}, height={}",
                width, height
            ))
        } else {
            None
        }
    }

    fn videocrop(&self) -> Option<String> {
        if let Some(ref coords) = self.builder.coordinates {
            let stream = &self.streams()[0];

            let actual_screen = self.builder.actual_screen.as_ref().unwrap();
            let stream_width = stream.size().unwrap().0;
            let stream_height = stream.size().unwrap().1;

            let scale_factor = (stream_width / actual_screen.width) as f64;
            let coords = coords.clone().rescale(scale_factor);

            let top_crop = coords.y;
            let left_crop = coords.x;
            let right_crop = stream_width as f64 - (coords.width + coords.x);
            let bottom_crop = stream_height as f64 - (coords.height + coords.y);

            // It is a requirement for x264enc to have even resolution.
            Some(format!(
                "videocrop top={} left={} right={} bottom={}",
                utils::round_to_even(top_crop),
                utils::round_to_even(left_crop),
                utils::round_to_even(right_crop),
                utils::round_to_even(bottom_crop)
            ))
        } else {
            None
        }
    }

    fn videoenc(&self) -> String {
        let value = env::var("GST_VAAPI_ALL_DRIVERS").unwrap_or_default();
        let is_use_vaapi = value == "1";
        log::debug!("is_use_vaapi: {}", is_use_vaapi);

        if is_use_vaapi {
            match self.video_format() {
                VideoFormat::Webm | VideoFormat::Mkv => "vaapivp8enc", // FIXME Improve pipelines
                VideoFormat::Mp4 => "vaapih264enc ! h264parse",
                VideoFormat::Gif => "gifenc speed=30", // FIXME This doesn't really use vaapi
            }
        } else {
            match self.video_format() {
                VideoFormat::Webm | VideoFormat::Mkv => "vp8enc max_quantizer=17 cpu-used=16 cq_level=13 deadline=1 static-threshold=100 keyframe-mode=disabled buffer-size=20000 threads=%T",
                VideoFormat::Mp4 => "x264enc qp-max=17 speed-preset=superfast threads=%T ! video/x-h264, profile=baseline",
                VideoFormat::Gif => "gifenc speed=30",
            }
        }.to_string()
    }

    fn audioenc(&self) -> Option<String> {
        match self.video_format() {
            VideoFormat::Webm | VideoFormat::Mkv | VideoFormat::Mp4 => Some("opusenc"),
            VideoFormat::Gif => None,
        }
        .map(str::to_string)
    }

    fn muxer(&self) -> Option<String> {
        let video_format = self.video_format();

        let muxer = match video_format {
            VideoFormat::Webm => "webmmux",
            VideoFormat::Mkv => "matroskamux",
            VideoFormat::Mp4 => "mp4mux",
            VideoFormat::Gif => return None,
        };

        Some(format!("{} name=mux", muxer))
    }

    fn video_format(&self) -> VideoFormat {
        let file_extension = self.file_path().extension().unwrap().to_str().unwrap();

        VideoFormat::from_str(file_extension).expect("Invalid video format.")
    }

    fn framerate(&self) -> u32 {
        if self.video_format() == VideoFormat::Gif {
            return GIF_DEFAULT_FRAMERATE;
        }

        self.builder.framerate
    }

    fn file_path(&self) -> &Path {
        self.builder.file_path.as_path()
    }

    fn speaker_source(&self) -> Option<&str> {
        self.builder.speaker_source.as_deref()
    }

    fn mic_source(&self) -> Option<&str> {
        self.builder.mic_source.as_deref()
    }

    fn fd(&self) -> i32 {
        self.builder.fd
    }

    fn streams(&self) -> &Vec<Stream> {
        &self.builder.streams
    }

    fn is_single_stream(&self) -> bool {
        self.streams().len() == 1
    }
}
